mods/taskbar-icon-numberer.wh.cpp

```diff
@@ -0,0 +1,468 @@
+                break;
+            default: // bottomRight
+                textContainer.HorizontalAlignment(HorizontalAlignment::Right);
+                textContainer.VerticalAlignment(VerticalAlignment::Bottom);
+                break;
+        }
+        textContainer.Margin(Thickness{0, 0, 2, 2});
+        
+        Canvas::SetZIndex(textContainer, 1000);
+        
+        std::lock_guard<std::mutex> lock(g_overlayMutex);
+        iconPanelElement.as<Panel>().Children().Append(textContainer);
+        g_numberedButtons.insert(buttonPtr);
+        
+    } catch (...) {
+        // Ignore errors
+    }
+}
+
+void ClearAllOverlays() {
+    std::lock_guard<std::mutex> lock(g_overlayMutex);
+    g_numberedButtons.clear();
+}
+
+void UpdateAllTaskbarNumbers(FrameworkElement taskbarRepeater) {
+    if (g_unloading) return;
+    
+    try {
+        auto panel = taskbarRepeater.as<Panel>();
+        if (!panel) return;
+        
+        auto children = panel.Children();
+        
+        // Collect all TaskListButtons with their X positions
+        std::vector<std::pair<double, FrameworkElement>> buttons;
+        
+        for (uint32_t i = 0; i < children.Size(); i++) {
+            auto child = children.GetAt(i).try_as<FrameworkElement>();
+            if (!child || child.Name() != L"TaskListButton") continue;
+            
+            double x = child.ActualOffset().x;
+            buttons.push_back({x, child});
+        }
+        
+        // Sort by X position (left to right)
+        std::sort(buttons.begin(), buttons.end(), 
+                  [](const auto& a, const auto& b) { return a.first < b.first; });
+        
+        // Number buttons in visual order
+        int buttonIndex = 0;
+        for (const auto& [x, button] : buttons) {
+            if (buttonIndex >= g_settings.maxNumbers) break;
+            
+            bool shouldShow = true;
+            if (g_settings.showOnlyRunning) {
+                shouldShow = TaskListButton_IsRunning(button);
+            }
+            
+            if (shouldShow) {
+                buttonIndex++;
+                CreateNumberOverlay(button, buttonIndex);
+            } else {
+                // Remove overlay
+                auto iconPanel = FindChildByName(button, L"IconPanel");
+                if (iconPanel) {
+                    auto existingOverlay = FindChildByName(iconPanel, L"WindhawkNumberOverlay");
+                    if (existingOverlay) {
+                        auto panelChildren = iconPanel.as<Panel>().Children();
+                        uint32_t index;
+                        if (panelChildren.IndexOf(existingOverlay, index)) {
+                            panelChildren.RemoveAt(index);
+                        }
+                    }
+                }
+            }
+        }
+        
+    } catch (...) {
+        // Ignore errors
+    }
+}
+
+void WINAPI TaskListButton_UpdateVisualStates_Hook(void* pThis) {
```

Comment:
Note that this function is called for secondary taskbars too, so numbers are added on all monitors. It makes more sense for the mod to only show numbers on the primary taskbar since that's what Win+# hotkeys affect.

Reply:
In one of the mods, I'm using this method, it isn't very elegant but it's been working correctly so far. 

https://github.com/ramensoftware/windhawk-mods/blob/54dde4e2305ee64046cb84ffb2520f83dfe26f4f/mods/taskbar-vertical.wh.cpp#L1363-L1381